keps/sig-auth/5681-conditional-authorization/README.md

```diff
@@ -0,0 +1,1425 @@
+
+    // Conditions is an unordered set of conditions that should be evaluated
+    // against admission attributes, to determine
+    // whether this authorizer allows the request.
+    //
+    // +listType=map
+    // +listMapKey=id
+    // +optional
+    Conditions []SubjectAccessReviewCondition `json:"conditions,omitempty"`
+}
+
+type SubjectAccessReviewCondition struct {
+    ID string                                       `json:"id"`
+    Effect      SubjectAccessReviewConditionEffect  `json:"effect"`
+    Type        string                              `json:"type"`
+    Condition   string                              `json:"condition"`
+    Description string                              `json:"description,omitempty"`
+}
+
+type SubjectAccessReviewConditionEffect string
+
+const (
+    SubjectAccessReviewConditionEffectAllow     SubjectAccessReviewConditionEffect = "Allow"
+    SubjectAccessReviewConditionEffectDeny      SubjectAccessReviewConditionEffect = "Deny"
+    SubjectAccessReviewConditionEffectNoOpinion SubjectAccessReviewConditionEffect = "NoOpinion"
+)
+```
+
+`Status.ConditionsChain` is mutually exclusive with `Status.Allowed` and
+`Status.Denied`. A conditional response is characterized by
+`Status.ConditionsChain != null`. Old implementers that do not recognize
+`Status.ConditionsChain` will just safely assume it was a `NoOpinion`.
+
+### Changes to the Webhook Authorizer
+
+Thanks to the `SubjectAccessReviewStatus` API change, the webhook authorizer
+supports conditional responses in a straightforward manner. The webhook checks
+if the feature is enabled, `Status.ConditionsChain` is set, there is only one
+ConditionSet returned (there is a 1:1 mapping between an authorizer and a
+ConditionSet), that conditional authorization is configured (as below), as per
+the following code snippet:
+
+```go
+gotConditionalResponse := len(r.Status.ConditionsChain) != 0 && utilfeature.DefaultFeatureGate.Enabled(genericfeatures.ConditionalAuthorization)
+
+switch {
+case r.Status.Denied && r.Status.Allowed:
+    return authorizer.DecisionDeny, r.Status.Reason, fmt.Errorf("webhook subject access review returned both allow and deny response")
+case r.Status.Denied && gotConditionalResponse:
+    return authorizer.DecisionDeny, r.Status.Reason, fmt.Errorf("webhook subject access review returned both deny and conditional response")
+case r.Status.Allowed && gotConditionalResponse:
+    return w.decisionOnError, r.Status.Reason, fmt.Errorf("webhook subject access review returned both allow and conditional response")
+case r.Status.Denied:
+    return authorizer.DecisionDeny, r.Status.Reason, nil
+case r.Status.Allowed:
+    return authorizer.DecisionAllow, r.Status.Reason, nil
+case gotConditionalResponse:
+    if !w.supportsConditionalResponse {
+        return w.decisionOnError, "", fmt.Errorf("webhook authorizer does not support conditional responses") 
+    }
+
+    if len(r.Status.ConditionsChain) > 1 {
```

Comment:
Relax this to allow any amount of chained authorizers, this is needed when an aggregated API server asks kube-apiserver whether to authorize a request; kube-apiserver might return multiple `ConditionSet`s, and the aggregated API server should evaluate and respect those condition sets in order.